### PR TITLE
Fms package update

### PR DESCRIPTION
**Pull request by rem1776**
adds more versions of FMS and more CMake build options, plus some missing build dependencies

### CODE REVIEW

**Comment by kgerheiser** [comment]
Looks good to me

**Review by kgerheiser — DISMISSED** [review]

File: var/spack/repos/builtin/packages/fms/package.py [hunk 1]
```diff
@@ -13,18 +13,33 @@ class Fms(CMakePackage):
...
+    url      = "https://github.com/NOAA-GFDL/FMS/archive/refs/tags/2021.04.tar.gz"
+    git      = "https://github.com/NOAA-GFDL/FMS.git"
 
     maintainers = ['kgerheiser', 'Hang-Lei-NOAA', 'edwardhartnett']
 
     variant('64bit', default=True, description='64 bit?')
     variant('gfs_phys', default=True, description='Use GFS Physics?')
     variant('openmp', default=True, description='Use OpenMP?')
+    variant('enable_quad_precision', default=True, description='quad precision reals?')
+    variant('with_yaml', default=True, description='yaml input file support(requires libyaml)?')
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/fms/package.py]: Can we avoid `with_*` or `enable_*`? These should be implied by `+` in the variant:
```suggestion
    variant('quad_precision', default=True, description='quad precision reals')
    variant('yaml', default=True, description='yaml input file support')
```
Also, we shouldn't have any question marks in description (users looking at Spack info might be confused by reading `(requires libyaml)?`)

**Review by alalazo — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/fms/package.py (hunk 1 above)
Comment (rem1776) [reply var/spack/repos/builtin/packages/fms/package.py]: @alalazo It's updated now with those changes

File: var/spack/repos/builtin/packages/fms/package.py [hunk 2]
```diff
@@ -13,19 +13,34 @@ class Fms(CMakePackage):
...
 
-    variant('64bit', default=True, description='64 bit?')
-    variant('gfs_phys', default=True, description='Use GFS Physics?')
-    variant('openmp', default=True, description='Use OpenMP?')
+    variant('64bit', default=True, description='64 bit')
+    variant('gfs_phys', default=True, description='Use GFS Physics')
+    variant('openmp', default=True, description='Use OpenMP')
+    variant('quad_precision', default=True, description='quad precision reals')
+    variant('yaml', default=False, description='yaml input file support(requires libyaml)')
+    variant('constants', values=str, default='GFDL', description='Build with <X> constants parameter definitions')
```
Comment (kgerheiser) [inline var/spack/repos/builtin/packages/fms/package.py]: These options need equivalent CMake args. If `yaml`, for example, is enabled in Spack that option is never passed to CMake.

**Review by kgerheiser — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/fms/package.py [hunk 3]
```diff
@@ -13,25 +13,45 @@ class Fms(CMakePackage):
...
     version('2021.03.01', sha256='1f70e2a57f0d01e80fceb9ca9ce9661f5c1565d0437ab67618c2c4dfea0da6e9')
-
+    version('2021.03',    sha256='a9fb6e85f44ff79e6f9e61e65f42a5ffd38fa661fe1a3e4da6f85ffacd2697ac')
+    version('2021.02.01', sha256='9b11d9474d7c90464af66d81fb86c4798cfa309b9a0da20b0fccf33c4f65386b')
+    version('2021.02',    sha256='db810b2452a6952239f064b52c0c5c58fc62126057982111b9fcd64f1b3bd879')
+    version('2021.01',    sha256='38c748e2edb94ffeb021095d8bde4d74b7834610ce0ef1dbb4dce353eeb5cd96')
+    version('2020.04.02', sha256='bd6ce752b1018d4418398f14b9fc486f217de76bcbaaf2cdbf4c43e0b3f39f69')
+    version('2020.04.01', sha256='2c409242de7dea0cf29f8dbf7495698b6bcac1eeb5c4599a728bdea172ffe37c')
+
+    depends_on('cmake')
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/fms/package.py]: This is implied by the base class:
```suggestion
```

File: var/spack/repos/builtin/packages/fms/package.py [hunk 4]
```diff
@@ -13,25 +13,45 @@ class Fms(CMakePackage):
...
+
+    variant('64bit', default=True, description='Build a version of the library with default 64 bit reals')
+    variant('gfs_phys', default=True, description='Use GFS Physics')
+    variant('openmp', default=True, description='Use OpenMP')
+    variant('quad_precision', default=True, description='quad precision reals')
+    variant('yaml', default=False, description='yaml input file support(requires libyaml)', when='@2021.04')
+    variant('constants', default='GFDL', description='Build with <X> constants parameter definitions',
+            values=('GFDL','GEOS','GFS'), multi=False, when='@2022.02')
+    variant('fpic', default=False, description='Build with position independent code', when='@2022.02')
+    version('2022.02',    sha256='ad4978302b219e11b883b2f52519e1ee455137ad947474abb316c8654f72c874')
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/fms/package.py]: It's common practice in all other packages to have versions before variants. Can we do the same here?

File: var/spack/repos/builtin/packages/fms/package.py (hunk 2 above)
Comment (rem1776) [reply var/spack/repos/builtin/packages/fms/package.py]: The cmake args for those variants are added with 1710637 along with some other improvements to the options

File: var/spack/repos/builtin/packages/fms/package.py [hunk 5]
```diff
@@ -13,25 +13,45 @@ class Fms(CMakePackage):
...
+
+    maintainers = ['kgerheiser', 'Hang-Lei-NOAA', 'edwardhartnett', 'rem1776']
+
+    variant('64bit', default=True, description='Build a version of the library with default 64 bit reals')
+    variant('gfs_phys', default=True, description='Use GFS Physics')
+    variant('openmp', default=True, description='Use OpenMP')
+    variant('quad_precision', default=True, description='quad precision reals')
+    variant('yaml', default=False, description='yaml input file support(requires libyaml)', when='@2021.04')
+    variant('constants', default='GFDL', description='Build with <X> constants parameter definitions',
+            values=('GFDL','GEOS','GFS'), multi=False, when='@2022.02')
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/fms/package.py]: Maybe you meant:
```suggestion
            values=('GFDL','GEOS','GFS'), multi=False, when='@2022.02:')
```
? (The option will not disappear after 2022.02)

**Review by alalazo — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/fms/package.py (hunk 5 above)
Comment (rem1776) [reply var/spack/repos/builtin/packages/fms/package.py]: I had thought specifying the version this way automatically worked with any more recent versions, does it need the `:` as well?

Comment (alalazo) [reply var/spack/repos/builtin/packages/fms/package.py]: Yes

Comment (rem1776) [reply var/spack/repos/builtin/packages/fms/package.py]: Thanks for pointing that out, all the comments should be addressed now

File: var/spack/repos/builtin/packages/fms/package.py [hunk 6]
```diff
@@ -13,25 +13,45 @@ class Fms(CMakePackage):
...
 
     def cmake_args(self):
         args = [
             self.define_from_variant('64BIT'),
             self.define_from_variant('GFS_PHYS'),
-            self.define_from_variant('OPENMP')
+            self.define_from_variant('OPENMP'),
+            self.define_from_variant('ENABLE_QUAD_PRECISION', 'quad_precision'),
+            self.define_from_variant('WITH_YAML', 'yaml'),
+            self.define_from_variant('CONSTANTS')
```
Comment (kgerheiser) [inline var/spack/repos/builtin/packages/fms/package.py]: Missing a comma

**Review by kgerheiser — APPROVED** [review]

**Review by alalazo — APPROVED** [review]

**Pull request merged**